### PR TITLE
doClientKeyExchange improvements

### PR DESCRIPTION
**Pull request by rizlik**
# Description
Improve doClientKeyExchange code

# Testing
`./configure --enable-all && make check`

### CODE REVIEW

**Comment by dgarske** [comment]
@SparkiDev and @kaleb-himes

I am working on fixing the Jenkins error. I can reproduce only on Ubuntu using:

```
./configure --enable-dtls --enable-dtls-mtu --enable-maxfragment --enable-secure-renegotiation CFLAGS="-DWOLFSSL_STATIC_RSA -DWOLFSSL_STATIC_DH" CC="clang -fsanitize=address"  && make
./examples/server/server -v 2 -l AES128-SHA -2 &
./examples/client/client -v 2 -l AES128-SHA -2

received record layer msg
got CHANGE CIPHER SPEC
growing input buffer
Verify MAC of Encrypted Data
VerifyMacEnc failed
wolfSSL error occurred, error = -305
wolfSSL error occurred, error = -312
```

File: src/internal.c [hunk 1]
```diff
@@ -4672,11 +4677,13 @@ int RsaDec(WOLFSSL* ssl, byte* in, word32 inSz, byte** out, word32* outSz,
...
-        *outSz = ret;
-        ret = 0;
-    }
+    mask = ctMaskGT(ret, 0);
+    *outSz = 0;
+    zero = 0;
+    outSzTmp = (word32)ret;
+    ctMaskCopy(mask, (byte*)outSz, (byte*)&outSzTmp, (byte*)outSz, sizeof(*outSz));
+    ctMaskCopy(mask, (byte*)&ret, (byte*)&zero, (byte*)&ret, sizeof(ret));
+    ctMaskCopy(mask, (byte*)out, (byte*)&outTmp, (byte*)out, sizeof(*out));
```
Comment (SparkiDev) [inline src/internal.c:4686]: Prefer the ctMaskCopy copied over existing buffer when mask is set.
Then you can use:
    r[i] ^= (r[i] ^ a[i]) & mask;

File: src/internal.c [hunk 2]
```diff
@@ -4672,11 +4677,13 @@ int RsaDec(WOLFSSL* ssl, byte* in, word32 inSz, byte** out, word32* outSz,
...
-    /* For positive response return in outSz */
-    if (ret > 0) {
-        *outSz = ret;
-        ret = 0;
-    }
+    mask = ctMaskGT(ret, 0);
+    *outSz = 0;
+    zero = 0;
+    outSzTmp = (word32)ret;
+    ctMaskCopy(mask, (byte*)outSz, (byte*)&outSzTmp, (byte*)outSz, sizeof(*outSz));
```
Comment (SparkiDev) [inline src/internal.c:4684]: *outSz ^= (*outSz ^ outSzTmp) & mask;

File: src/internal.c [hunk 3]
```diff
@@ -4672,11 +4677,13 @@ int RsaDec(WOLFSSL* ssl, byte* in, word32 inSz, byte** out, word32* outSz,
...
+    ctMaskCopy(mask, (byte*)&ret, (byte*)&zero, (byte*)&ret, sizeof(ret));
```
Comment (SparkiDev) [inline src/internal.c:4685]: ret &= mask

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by dgarske** [comment]
@SparkiDev , since Marco is out can you please refactor the code how you suggested?

**Review by SparkiDev — APPROVED** [review]

**Review by dgarske — APPROVED** [review]

**Pull request merged**